pkg/workloads/containerd/watcher.go

```diff
@@ -53,106 +52,80 @@ const (
 	eventQueueBufferSize = 100
 )
 
-// containerEvents holds per-container queues for events
-type containerEvents struct {
-	lock.Mutex
-	events map[string]chan dTypesEvents.Message
-}
-
-func (ce *containerEvents) enqueueByContainerID(e dTypesEvents.Message) {
-	ce.Lock()
-	defer ce.Unlock()
-
-	if _, found := ce.events[e.Actor.ID]; !found {
-		q := make(chan dTypesEvents.Message, eventQueueBufferSize)
-		ce.events[e.Actor.ID] = q
-		go processContainerEvents(e.Actor.ID, q)
-	}
-	ce.events[e.Actor.ID] <- e
-}
-
-func (ce *containerEvents) reapEmpty() {
-	ce.Lock()
-	defer ce.Unlock()
-
-	for id, q := range ce.events {
-		if len(q) == 0 {
-			close(q)
-			delete(ce.events, id)
-		}
-	}
-}
-
 func shortContainerID(id string) string {
 	return id[:10]
 }
 
 // EnableEventListener watches for docker events. Performs the plumbing for the
 // containers started or dead.
 func EnableEventListener() error {
-	eventQueue := containerEvents{events: make(map[string]chan dTypesEvents.Message)}
-	since := time.Now()
-	syncWithRuntime()
+	ws := newWatcherState(eventQueueBufferSize)
+	syncWithRuntime(ws)
 
+	since := time.Now()
 	eo := dTypes.EventsOptions{Since: strconv.FormatInt(since.Unix(), 10)}
 	r, err := dockerClient.Events(ctx.Background(), eo)
 	if err != nil {
 		return err
 	}
 
-	go listenForDockerEvents(&eventQueue, r)
+	go listenForDockerEvents(ws, r)
 
 	// start a go routine which periodically synchronizes containers
 	// managed by the local container runtime and checks if any of them
 	// need to be managed by Cilium. This is a fall back mechanism in case
 	// an event notification has been lost.
-	go func() {
+	go func(state *watcherState) {
 		for {
 			time.Sleep(syncRateDocker)
-			syncWithRuntime()
 
-			log.Debug("Reaping empty event queues")
-			eventQueue.reapEmpty()
+			log.Debug("Reaping empty containerd watcher event queues")
+			state.reapEmpty()
+
+			log.Debug("Periodic watcher synchronization to containerd runtime")
+			syncWithRuntime(state)
 		}
-	}()
+	}(ws)
 
 	log.Debugf("Started to listen for containerd events")
-
 	return nil
 }
 
 // syncWithRuntime is used by the daemon to synchronize changes between Docker and
 // Cilium. This includes identities, labels, etc.
-func syncWithRuntime() {
+func syncWithRuntime(ws *watcherState) {
 	var wg sync.WaitGroup
 
-	// FIXME GH-1662: Must be synchronize with event handler
-
 	cList, err := dockerClient.ContainerList(ctx.Background(), dTypes.ContainerListOptions{All: false})
 	if err != nil {
 		log.Errorf("Failed to retrieve the container list %s", err)
+		return
```

Comment:
This is new behaviour (it wasn't there for the first review). If `err != nil` I'm pretty sure we can't do much with cList (ideally it's nil but it might be trash).